python/neuroglancer/pipeline/volumes/cloudvolume.py

```diff
@@ -0,0 +1,486 @@
+
+    return info
+
+  @classmethod
+  def extract_path(cls, cloudpath):
+    """cloudpath: e.g. gs://neuroglancer/DATASET/LAYER/info or s3://..."""
+    match = re.match(r'^(gs|file|s3)://([/\d\w_\.\-]+)/([\d\w_\.\-]+)/([\d\w_\.\-]+)/?(?:info)?', cloudpath)
+    return ExtractedPath(*match.groups())
+
+  def refreshInfo(self):
+    infojson = self._storage.get_file('info')
+    self.info = json.loads(infojson)
+    return self.info
+
+  def commitInfo(self):
+    infojson = json.dumps(self.info)
+    self._storage.put_file('info', infojson, 'application/json').wait()
+    return self
+
+  @property
+  def dataset_name(self):
+    return self._dataset_name
+
+  @dataset_name.setter
+  def dataset_name(self, name):
+    if name != self._dataset_name:
+      self._dataset_name = name
+      self.refreshInfo()
+  
+  @property
+  def layer(self):
+    return self._layer
+
+  @layer.setter
+  def layer(self, name):
+    if name != self._layer:
+      self._layer = name
+      self.refreshInfo()
+
+  @property
+  def scales(self):
+    return self.info['scales']
+
+  @property
+  def scale(self):
+    return self.mip_scale(self.mip)
+
+  def mip_scale(self, mip):
+    return self.info['scales'][mip]
+
+  @property
+  def base_cloudpath(self):
+    return "{}://{}/{}/".format(self._protocol, self._bucket, self.dataset_name)
+
+  @property
+  def layer_cloudpath(self):
+    return os.path.join(self.base_cloudpath, self.layer)
+
+  @property
+  def info_cloudpath(self):
+    return os.path.join(self.layer_cloudpath, 'info')
+
+  @property
+  def shape(self):
+    return self.mip_shape(self.mip)
+
+  def mip_shape(self, mip):
+    size = self.mip_volume_size(mip)
+    return Vec(size.x, size.y, size.z, self.num_channels)
+
+  @property
+  def volume_size(self):
+    return self.mip_volume_size(self.mip)
+
+  def mip_volume_size(self, mip):
+    return Vec(*self.info['scales'][mip]['size'])
+
+  @property
+  def available_mips(self):
+    return range(len(self.info['scales']))
+
+  @property
+  def layer_type(self):
+    return self.info['type']
+
+  @property
+  def dtype(self):
+    return self.data_type
+
+  @property
+  def data_type(self):
+    return self.info['data_type']
+
+  @property
+  def encoding(self):
+    return self.mip_encoding(self.mip)
+
+  def mip_encoding(self, mip):
+    return self.info['scales'][mip]['encoding']
+
+  @property
+  def num_channels(self):
+    return int(self.info['num_channels'])
+
+  @property
+  def voxel_offset(self):
+    return self.mip_voxel_offset(self.mip)
+
+  def mip_voxel_offset(self, mip):
+    return Vec(*self.info['scales'][mip]['voxel_offset'])
+
+  @property 
+  def resolution(self):
+    return self.mip_resolution(self.mip)
+
+  def mip_resolution(self, mip):
+    return Vec(*self.info['scales'][mip]['resolution'])
+
+  @property
+  def downsample_ratio(self):
+    return self.resolution / self.mip_resolution(0)
+
+  @property
+  def underlying(self):
```

Comment:
this is a very confusing term to use. we should call this `underlying_chunk_size` or something.